src/canard.c

```diff
@@ -1,6 +1,736 @@
+
+  // // Resolving the state flags:
+  const bool not_initialized = (rxstate->timestamp_usec == 0) ? true : false;
+  // const bool tid_timed_out = (rxstate->timestamp_usec - timestamp_usec) > TRANSFER_TIMEOUT_USEC;
+  const bool first_frame = IS_START_OF_TRANSFER(tail_byte);
+  const bool not_previous_tid = computeForwardDistance(rxstate->transfer_id, TRANSFER_ID_FROM_TAIL_BYTE(tail_byte)) > 1;
+
+  bool need_restart = 
+        (not_initialized) ||
+        // (tid_timed_out) ||
+        (first_frame && not_previous_tid);
+
+  if (need_restart)
+  {
+    rxstate->transfer_id = TRANSFER_ID_FROM_TAIL_BYTE(tail_byte);
+    rxstate->next_toggle = 0;
+    canardReleaseStatePayload(ins, rxstate);
+    if (!IS_START_OF_TRANSFER(tail_byte))
+    {
+      rxstate->transfer_id += 1;
+      if (rxstate->transfer_id >= 32)
+      {
+        rxstate->transfer_id = 0;
+      }
+      return;
+    }
+  }
+
+  if (IS_START_OF_TRANSFER(tail_byte) && IS_END_OF_TRANSFER(tail_byte))
+  {  // single frame transfer
+
+    rxstate->timestamp_usec = timestamp_usec;
+    CanardRxTransfer rxtransfer = {
+    .payload_head = frame->data,
+    .payload_len = frame->data_len-1,
+    .data_type_id = data_type_id,
+    .transfer_type = transfer_type,
+    .priority = priority,
+    .source_node_id = source_node_id };
+    ins->on_reception(ins,&rxtransfer);
+
+    canardPrepareForNextTransfer(rxstate);
+    return;
+
+  }
+
+  if (TOGGLE_BIT(tail_byte) != rxstate->next_toggle)
+  {
+    return;
+  }
+
+  if (TRANSFER_ID_FROM_TAIL_BYTE(tail_byte) != rxstate->transfer_id)
+  {
+    return;
+  }
+
+  if (IS_START_OF_TRANSFER(tail_byte) && !IS_END_OF_TRANSFER(tail_byte))
+  {  // beginning of multi frame transfer
+    //take off the crc and store the payload
+    rxstate->timestamp_usec = timestamp_usec;
+    canardBufferBlockPushBytes(&ins->allocator, rxstate, frame->data+2, frame->data_len-3);
+
+  }
+  else if (!IS_START_OF_TRANSFER(tail_byte) && !IS_END_OF_TRANSFER(tail_byte))
+  {
+
+    canardBufferBlockPushBytes(&ins->allocator, rxstate, frame->data, frame->data_len-1);
+
+  }
+  else
+  { 
+    CanardRxTransfer rxtransfer = {
+    .timestamp_usec = timestamp_usec,
+    .payload_head = rxstate->buffer_head,
+    .payload_middle = rxstate->buffer_blocks,
+    .payload_tail = frame->data,
+    .middle_len = rxstate->payload_len - CANARD_RX_PAYLOAD_HEAD_SIZE,
+    .payload_len = rxstate->payload_len+frame->data_len-1,
+    .data_type_id = data_type_id,
+    .transfer_type = transfer_type,
+    .priority = priority,
+    .source_node_id = source_node_id };
+
+    //crc validation goes here!
+
+    ins->on_reception(ins,&rxtransfer);
+
+    canardPrepareForNextTransfer(rxstate);
+
+    return;
+  }
+  rxstate->next_toggle ^= 1;
+}
+
+/**
+ * Traverses the list of transfers and removes those that were last updated more than 
+ * timeout_usec microseconds ago
+ */
+void canardCleanupStaleTransfers(CanardInstance* ins, uint64_t timeout_usec, uint64_t current_time_usec)
```

Comment:
When we were working on the design document, the specification has not had a notion of *transfer ID timeout*. Now, with the concept of transfer ID timeout explicitly specified, we do not have to expose the `timeout_usec` to the application. Instead we can just hard-code it as 2 seconds. Relevant part of the specification: http://uavcan.org/Specification/4._CAN_bus_transport_layer/#transfer-id-computation